pulsar-client/src/main/java/org/apache/pulsar/client/impl/ConsumerImpl.java

```diff
@@ -529,6 +534,51 @@ protected CompletableFuture<Messages<T>> internalBatchReceiveAsync() {
         return result;
     }
 
+    private void processMessageIdBeforeAcknowledge(MessageIdImpl messageId, AckType ackType, int numMessages) {
+        if (ackType == AckType.Individual) {
+            stats.incrementNumAcksSent(numMessages);
+            unAckedMessageTracker.remove(messageId);
+            if (possibleSendToDeadLetterTopicMessages != null) {
+                possibleSendToDeadLetterTopicMessages.remove(messageId);
+            }
+        } else {
+            stats.incrementNumAcksSent(unAckedMessageTracker.removeMessagesTill(messageId));
+        }
+    }
+
+    private @Nullable MessageIdImpl getMessageIdToAcknowledge(BatchMessageIdImpl messageId, AckType ackType) {
+        final BatchMessageAcker acker;
+        if (messageId.getAcker() instanceof BatchMessageAckerDisabled) {
+            acker = batchMessageToAcker.computeIfAbsent(
+                    Pair.of(messageId.getLedgerId(), messageId.getEntryId()),
+                    __ -> BatchMessageAcker.newAcker(messageId.getBatchSize()));
```

Comment:
Possible memory leak here. E.g:

```java
// deserialized message id {batchSize = 2, batchIndex = 1}
// and { batchIndex = 0 } already has been acked before.
MessageId msgId = MessageId.fromByteArray(msgId.toByteArray());
consumer.acknowledge(msgId);
// then the acker in the map will no longer be released.
```

Reply:
It's hard to handle this case if you meant some `MessageId`s are created from the `MessageId` and other `MessageId`s are created from `fromByteArray`.

This PR assumes that when users acknowledge a `MessageId` that is created from `fromByteArray`, the other `MessageId` instances to acknowledge are also created from `fromByteArray`. I'm wondering in which case would the user adopt different ways to create `MessageId`.

```java
var msg = consumer.receive(); // batch index: 0, batch size: 2
consumer.acknowledge(msg);
// Why not acknowledge the next msg, 
var msgId = MessageId.fromByteArray(readMessageIdFromDataBase()); // batch index: 1, batch size: 2
consumer.acknowledge(msgId);
```

Reply:
If all clients are not enabled `BatchMessageAcker`, this will not happen